### ISSUE TITLE
Importing MySQL data from a .sql file

### ISSUE THREAD

**Issue by spruce-bruce** [tgriesser/knex#944]
I have an existing MySQL database that I have to work with. In development I want to use knex seeds to create this database. I don't HAVE to do this w/ knex (I can use chef if I need to), but it'd be nice to be able to do.

I create a .sql file with the command `mysql -u user -p table > dump.sql`. The output contains conditional execution statements that look like this:
```
/*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */;
/*!40101 SET @OLD_CHARACTER_SET_RESULTS=@@CHARACTER_SET_RESULTS */;
/*!40101 SET @OLD_COLLATION_CONNECTION=@@COLLATION_CONNECTION */;
/*!40101 SET NAMES utf8 */;
/*!40103 SET @OLD_TIME_ZONE=@@TIME_ZONE */;
/*!40103 SET TIME_ZONE='+00:00' */;
/*!40014 SET @OLD_UNIQUE_CHECKS=@@UNIQUE_CHECKS, UNIQUE_CHECKS=0 */;
/*!40014 SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0 */;
/*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
```

Knex barfs on these statements. My error output from `knex seed:run` is 
```
rror: ER_PARSE_ERROR: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set' at line 9
    at Query.Sequence._packetToError (/vagrant/node_modules/mysql/lib/protocol/sequences/Sequence.js:48:14)
    at Query.ErrorPacket (/vagrant/node_modules/mysql/lib/protocol/sequences/Query.js:83:18)
    at Protocol._parsePacket (/vagrant/node_modules/mysql/lib/protocol/Protocol.js:274:23)
    at Parser.write (/vagrant/node_modules/mysql/lib/protocol/Parser.js:77:12)
    at Protocol.write (/vagrant/node_modules/mysql/lib/protocol/Protocol.js:39:16)
    at Socket.<anonymous> (/vagrant/node_modules/mysql/lib/Connection.js:96:28)
    at Socket.EventEmitter.emit (events.js:95:17)
    at Socket.<anonymous> (_stream_readable.js:746:14)
    at Socket.EventEmitter.emit (events.js:92:17)
    at emitReadable_ (_stream_readable.js:408:10)
    at emitReadable (_stream_readable.js:404:5)
    at readableAddChunk (_stream_readable.js:165:9)
    at Socket.Readable.push (_stream_readable.js:127:10)
    at TCP.onread (net.js:526:21)
    at Protocol._enqueue (/vagrant/node_modules/mysql/lib/protocol/Protocol.js:135:48)
    at Connection.query (/vagrant/node_modules/mysql/lib/Connection.js:201:25)
    at /vagrant/node_modules/knex/lib/dialects/mysql/index.js:92:18
From previous event:
    at Client._query (/vagrant/node_modules/knex/lib/dialects/mysql/index.js:88:12)
    at Client.query (/vagrant/node_modules/knex/lib/client.js:127:24)
    at Runner.assign.query (/vagrant/node_modules/knex/lib/runner.js:118:24)
From previous event:
    at /vagrant/node_modules/knex/lib/runner.js:44:21
From previous event:
    at Runner.run (/vagrant/node_modules/knex/lib/runner.js:30:20)
    at Raw.Target.then (/vagrant/node_modules/knex/lib/interface.js:27:43)
    at Object.exports.seed (/vagrant/workbench-seeds/dev/workbench.js:5:20)
    at Object.<anonymous> (/vagrant/node_modules/knex/lib/seed/index.js:110:19)
From previous event:
    at /vagrant/node_modules/knex/lib/seed/index.js:109:23
    at arrayEach (/vagrant/node_modules/lodash/index.js:1289:13)
    at Function.<anonymous> (/vagrant/node_modules/lodash/index.js:3345:13)
    at Seeder._waterfallBatch (/vagrant/node_modules/knex/lib/seed/index.js:104:5)
    at Seeder.<anonymous> (/vagrant/node_modules/knex/lib/seed/index.js:64:19)
From previous event:
    at Seeder.<anonymous> (/vagrant/node_modules/knex/lib/seed/index.js:63:31)
From previous event:
    at Seeder._runSeeds (/vagrant/node_modules/knex/lib/seed/index.js:62:82)
    at Seeder.<anonymous> (/vagrant/node_modules/knex/lib/seed/index.js:24:17)
From previous event:
    at Seeder.<anonymous> (/vagrant/node_modules/knex/lib/seed/index.js:23:38)
From previous event:
    at Command.<anonymous> (/usr/local/lib/node_modules/knex/lib/bin/cli.js:134:34)
    at Command.listener (/usr/local/lib/node_modules/knex/node_modules/commander/index.js:301:8)
    at Command.EventEmitter.emit (events.js:98:17)
    at Command.parseArgs (/usr/local/lib/node_modules/knex/node_modules/commander/index.js:610:12)
    at Command.parse (/usr/local/lib/node_modules/knex/node_modules/commander/index.js:458:21)
    at Liftoff.invoke (/usr/local/lib/node_modules/knex/lib/bin/cli.js:142:13)
    at Liftoff.<anonymous> (/usr/local/lib/node_modules/knex/node_modules/liftoff/index.js:181:16)
    at module.exports (/usr/local/lib/node_modules/knex/node_modules/liftoff/node_modules/flagged-respawn/index.js:17:3)
    at Liftoff.<anonymous> (/usr/local/lib/node_modules/knex/node_modules/liftoff/index.js:174:9)
    at /usr/local/lib/node_modules/knex/node_modules/liftoff/index.js:148:9
    at /usr/local/lib/node_modules/knex/node_modules/v8flags/index.js:99:14
    at /usr/local/lib/node_modules/knex/node_modules/v8flags/index.js:38:7
    at process._tickCallback (node.js:415:13)
    at Function.Module.runMain (module.js:499:11)
    at startup (node.js:119:16)
    at node.js:902:3
```

And my seed file looks like this:
```
var fs = require('fs');

exports.seed = function(knex, Promise) {
    var sql = fs.readFileSync('./data/workbench.sql').toString();
    return Promise.join(
        knex.raw('DROP DATABASE workbench'),
        knex.raw('CREATE DATABASE workbench'),
        knex.raw(sql)
    );
};
```

Is there a better way to import a .sql file using knex? Is that not really what knex is for (ie, am I misusing the technology)? For now I'm going to have to work around this problem with chef, but I'm interested in what you guys think about this use case and problem.

**Comment by tolgaek** [comment]
I'm in the exact same boat minus the option to work with chef

**Comment by kevzettler** [comment]
I also have an existing .sql database that I would like to use as a seed a knex project. Or convert the existing schema into knex migrations. I'm not seeing a clear supported method for this

**Comment by dnascimento** [comment]
Use case: You defined a base schema in sql file and a set of dummy data for tests. 
In every test, you want to drop the database and load the schema and dummy data.

**Comment by ph3b** [comment]
Any updates on this?

**Comment by elhigu** [comment]
@ph3b nope... no updates

`<stating-the-obvious>`you can use external tool (mysql client etc.) to import base data and then write knex migrations to top of that and run them separately`</stating-the-obvious>`.

**Comment by ph3b** [comment]
I believe I found a solution. If you add `{ multipleStatements: true }` in the connection-object you pass the the Knex instance it will allow multipleStatements and not throw the parse error. The connection-object would then look something like:
```javascript
var knex = require('knex')({
  client: 'mysql',
  connection: {
    host     : '127.0.0.1',
    user     : 'your_database_user',
    password : 'your_database_password',
    database : 'myapp_test',
    multipleStatements: true
  }
});
```
Then you can do something similar to what @spruce-bruce is doing

```javascript
var fs = require('fs');

exports.seed = function(knex, Promise) {
    var sql = fs.readFileSync('./data/workbench.sql').toString();
    return knex.raw('DROP DATABASE workbench')
       .then(() => knex.raw('CREATE DATABASE workbench'))
       .then(() => knex.raw(sql))
};
```

**Comment by cavery8989** [comment]
`var productData = require('../../data/products/products.json');
exports.seed = function(knex, Promise) {
  // Deletes ALL existing entries
  return knex('products').del()
    .then(function () {
      return knex('products').insert(productData.data);
    });
};
`

**Comment by spruce-bruce** [comment]
@elhigu That is, in essence, what I did to work around the error I was getting.

We use vagrant machines for our development environment and the mysql dump file I was using was to set up a development database that closely matches one of our clients databases (including some test data).

We use vagrant and chef to create our development/qa/staging/production environments and I needed to set up this MySQL database in all but production.

I ended up setting up my development MySQL db using chef (which already creates the blank database anyway. It was just a matter of importing the mysql from the dump). This, ultimately, was not a problem, but I would have _preferred_ to do this with knex.

I'll have to give @ph3b's suggestion a try! But at this point my chef workaround has been fine for a year so it's certainly not critical.

**Comment by Burtonium** [comment]
How do you guys read the .sql file using knex?

**Comment by spruce-bruce** [comment]
I just read the file into memory and passed the contents to a raw knex query

**Comment by armenr** [comment]
During seeding, I want turn off foreign key checks for all inserts in the context of that seed file. 

Does anyone know how I can achieve this?

**Comment by MichelDiz** [comment]
There is something like this https://github.com/sequelize/sequelize-auto for Knex?
I'm migrating to knex from Sequelize, this "auto" tool helped a lot with my existing Mysql database.

best regards

**Comment by elhigu** [comment]
@MichelDiz no, there is no models nor validators / relations etc. in knex. You need to know your DB schema that you are using through knex.

**Comment by thu** [comment]
@spruce-bruce Here is my solution:
```javascript
const cp = require('child_process');

cp.exec('mysql -uyour_user -pyour_password < your_file.sql', (error, stdout, stderr) => {
    if (error) throw error;
    console.log(`stdout: ${stdout}`);
    console.log(`stderr: ${stderr}`);
});
```
I just post it on SO: [https://stackoverflow.com/a/47175173/3414180](https://stackoverflow.com/a/47175173/3414180)

**Comment by elhigu** [comment]
@azlan I don't think thats a good idea / also not really relevant since it is just trying to extract schema instead of importing SQL data dump.

**Comment by simeonpashley** [comment]
> ```js
> var fs = require('fs');
> 
> exports.seed = function(knex, Promise) {
>     var sql = fs.readFileSync('./data/workbench.sql').toString();
>     return knex.raw('DROP DATABASE workbench')
>        .then(() => knex.raw('CREATE DATABASE workbench'))
>        .then(() => knex.raw(sql))
> };
> ```

The [best working answer](https://github.com/knex/knex/issues/944#issuecomment-244346847) needed a slight tweak for me to fix runtime issues around relative path for SQL file and also a missing "USE" to set the database.

Apart from that, this solution is the best way to import a raw SQL file.

```js
var path = require('path');
var fs = require('fs');

exports.up = function (knex, Promise) {
  var sql = fs.readFileSync(path.join(__dirname, './data/workbench.sql')).toString();
  return knex
    .raw('DROP DATABASE workbench')
    .then(() => knex.raw('CREATE DATABASE workbench'))
    .then(() => knex.raw('USE workbench'))
    .then(() => knex.raw(sql));
};

exports.down = async function (knex, Promise) {};
```

**Comment by Brainicism** [comment]
For anyone else running using this solution and running into memory leaks, yep. `mysql` appears to consume the exact same amount of memory as the size of the SQL dump being imported, and never releasing it. Gave up and just ended up using execing the `mysql` CLI instead

**Comment by mercmobily** [comment]
This has been a long-standing request that has evolved significantly over the years through community discussion. This is a summary with some thoughts.

The current status is that Knex does not have native support for importing SQL dump files, but there are several working solutions that have emerged from the community. The issue remains open because even though workarounds exist, there's no official built-in method for this common use case.

Looking at the evolution of this issue, it started with a straightforward request to import MySQL data from .sql files, particularly for seeding and migration purposes. The core problem is that SQL dump files typically contain multiple statements, which Knex doesn't handle by default since it uses prepared statements for security reasons.

The most reliable solution that has emerged, as demonstrated by @ph3b in the discussion, involves enabling `multipleStatements: true` in the MySQL connection configuration. This allows you to read an SQL file and execute it through `knex.raw()`. The working code looks like this:

```javascript
var knex = require('knex')({
  client: 'mysql',
  connection: {
    host: '127.0.0.1',
    user: 'your_database_user',
    password: 'your_database_password',
    database: 'myapp_test',
    multipleStatements: true,
  },
});

var fs = require('fs');
var path = require('path');

exports.seed = function (knex, Promise) {
  var sql = fs
    .readFileSync(path.join(__dirname, './data/workbench.sql'))
    .toString();
  return knex
    .raw('DROP DATABASE workbench')
    .then(() => knex.raw('CREATE DATABASE workbench'))
    .then(() => knex.raw('USE workbench'))
    .then(() => knex.raw(sql));
};
```

However, there are significant concerns with this approach. As @Brainicism noted in the discussion, this method can cause memory leaks where the mysql driver consumes memory equal to the size of the SQL dump and never releases it. For large dumps, this becomes problematic. Additionally, enabling `multipleStatements` opens potential security vulnerabilities, particularly SQL injection risks, so it should only be used in controlled environments like test databases.

The alternative approach that several users have adopted is to use external tools. As @elhigu (a Knex maintainer) suggested, you can use the mysql client directly through child processes. @thu provided a working example:

```javascript
const cp = require('child_process');

cp.exec(
  'mysql -uyour_user -pyour_password < your_file.sql',
  (error, stdout, stderr) => {
    if (error) throw error;
    console.log(`stdout: ${stdout}`);
    console.log(`stderr: ${stderr}`);
  }
);
```

This approach avoids the memory issues and security concerns but requires the mysql client to be available in the environment.

What's still missing from Knex is:

1. **Native SQL import functionality** - A built-in method that safely handles SQL dump files without requiring `multipleStatements` or external tools
2. **Schema extraction tools** - The ability to generate Knex migrations from existing database schemas (similar to Sequelize's sequelize-auto)
3. **Proper statement parsing** - Support for executing multiple statements safely without security vulnerabilities
4. **Memory-efficient handling** - A solution that doesn't consume memory equal to the entire SQL dump size
5. **Database-agnostic approach** - Current workarounds are MySQL-specific; other databases may need different approaches

The maintainers' position, particularly @elhigu's comments, indicates that even though they understand the use case, implementing a complete SQL parser for this functionality would be complex and potentially unstable. The recommendation remains to use external tools for importing base data and then write Knex migrations on top of that.

For users encountering this issue today, the practical solutions are:

- Use the `multipleStatements: true` approach for small, controlled test environments
- Use external database client tools through child processes for production or large dumps
- Consider breaking down SQL dumps into individual Knex migrations for better control and version management

Related Issues:

- #257 - How to execute multiple queries in one run (closed, discusses the limitation of multiple statements)
- #581 - Set transaction isolation level (mentions multipleStatements)
- #1075 - Query Batch (closed, related to executing multiple queries)
- #6234 - Database Schema Introspection and Export (open, addresses schema extraction tools)
- #2573 - Memory leak when using transactions (open, related to memory issues)
- #1836 - stream() throwing "out of memory" errors (open, memory-related)
- #6254 - Unified Solution for Cross-Database RETURNING Inconsistencies (open, database-agnostic concerns)
- #5661 - How to make the following use-case database agnostic (open)
- #5994 - Help/Question: How to make the following use-case database agnostic (open)

**Comment by mercmobily** [comment]
I spent quite a while on this issue. I came up with 4 possible routes.  I need some feedback as  I am not sure which way we should lean. 9 years of discussion... and no real consensus!

So the options:

### Option 1: Native `knex.import()` Method

Add an official import method to Knex core with safety constraints:

```javascript
await knex.import('./dump.sql', {
  allowDestructive: true,  // Required for DROP/CREATE
  environment: 'test',      // Must match NODE_ENV
  streaming: true          // Process line-by-line to avoid memory issues
});
```

**Pros:**
- First-class support in Knex core
- Consistent API with other Knex methods
- Can leverage internal Knex knowledge for optimization

**Cons:**
- Increases core maintenance burden
- Security concerns need careful handling
- May bloat core for a feature not everyone needs

### Option 2: Official Plugin Package `@knex/import`

Create a separate, officially maintained package:

```javascript
const KnexImport = require('@knex/import');
const importer = new KnexImport(knex);
await importer.fromFile('./dump.sql');
```

**Pros:**
- Keeps core lean
- Can iterate faster than core
- Optional dependency
- Community can contribute more easily

**Cons:**
- Another package to maintain
- Potential version compatibility issues
- Not immediately discoverable

### Option 3: Documentation-Only Solution

Add comprehensive documentation and close this issue:

- Document the `multipleStatements: true` approach with security warnings
- Document the `child_process` approach as the recommended solution
- Provide complete working examples
- Link from migration guide

**Pros:**
- No new code to maintain
- Solutions already exist and work
- Immediate resolution

**Cons:**
- No progress on the actual feature request
- Users must implement their own error handling
- Memory issues remain unaddressed

### Option 4: Minimal Stream Parser

Implement a lightweight statement splitter that executes one statement at a time:

```javascript
await knex.migrate.importSQL('./dump.sql', {
  onProgress: (line, total) => console.log(`${line}/${total}`),
  stopOnError: true,
  batchSize: 1  // Execute one statement at a time
});
```

**Pros:**
- Solves memory issues
- No multipleStatements security risk
- No external dependencies
- Simple implementation (~200 lines)

**Cons:**
- Basic SQL parsing needed (quote awareness)
- Slower than multipleStatements approach
- Edge cases with complex SQL syntax

## Questions for the Community

1. **Which option would best solve your use case?** Please specify your database and typical dump file size.

2. **Is this primarily for development/testing or production use?** This greatly impacts security requirements.

3. **Would you prefer a solution that's slower but safer, or faster with documented risks?**

4. **For those using workarounds: What problems have you encountered?** Memory leaks? Security issues? Performance?

5. **Would you contribute to a `@knex/import` plugin if that path was chosen?**

## Additional Considerations

- Should this feature support all databases or start with MySQL/PostgreSQL?
- Should it handle compressed dumps (.sql.gz)?
- Is progress reporting important for your use case?
- Would you need transaction support (all-or-nothing imports)?

## Maintainer Input Needed

@elhigu @kibertoad @lorefnon - Your perspective on maintenance burden and security implications would be invaluable. Which approach aligns best with Knex's philosophy and roadmap?

## Next Steps

Based on feedback, we could:
1. Create a proof-of-concept for the most popular option
2. Start with documentation and see if that satisfies most needs
3. Form a working group to develop the plugin approach
4. Close this as out-of-scope if that's the consensus

Please let us know which options you support and comment with specific use cases or concerns. Let's find a path forward that works for everyone.

**Comment by Bessonov** [comment]
@mercmobily 

> However, there are significant concerns with this approach. As @Brainicism noted in the discussion, this method can cause memory leaks where the mysql driver consumes memory equal to the size of the SQL dump and never releases it. For large dumps, this becomes problematic. Additionally, enabling multipleStatements opens potential security vulnerabilities, particularly SQL injection risks, so it should only be used in controlled environments like test databases.

I'm using a different query builder, and my approach to migrations is to create a dedicated query builder instance specifically for that purpose. I pass raw SQL statements (which are bundled using a Webpack loader instead of `fs`) to the instance, then discard it once the migration is complete. This is necessary anyway due to differences in settings like timeouts, support for multiple statements, etc. I'm quite happy with this approach.

Wouldn't this method also help prevent memory leaks and improve security?

**Comment by gpetrov** [comment]
That will be useful although to preserve the full cross database idea behind knex, I would rather have the sql converted first to knex migration and then execute it.

**Comment by Bessonov** [comment]
Important context: I'm speaking about projects that are intended to use a specific database, not products that must support different databases at the same time.

After working in the software development field for around 25 years and migrating projects technologically and/or from one database to another, in _my_ experience, cross-database migration support never pays off.

First of all, it produces a lot of effort. To work with an SQL-based database, there is no way to avoid knowledge of SQL. But you won't use raw SQL, and therefore, for any more complicated use case, you are forced to look at examples and docs to figure out how to match the desired SQL with your query builder or ORM. This is error-prone, costs a lot of time, and sometimes you end up passing a raw query to your tool anyway because you haven't found "the tool way" to accomplish your task or you exceed the deadline while looking for the "right" solution.

Second, you don't care about it for migrations. If your tool supports cross-database migrations, that's great (but see the first point on that topic), yet you don't lose much if you use raw SQL. Using raw SQL is a lot easier, faster, and more powerful. You can even just try out your statements. If you switch your database, you are looking at dumping your data structure and data with tools like pg_dump. Deleting your old migrations and creating a fresh single "init" migration for the new database lets you benefit from a clean state and faster execution on CI.

Of course, there are some great ORMs like Django ORM or, to some extent, Spring Data JPA, where support for most use cases comes for free. That’s great and should be used.

Having said that, I would always opt in for a tool that supports multiple databases. This is because switching or learning a tool is a lot of effort. Switching a database happens very rarely, and if supporting this hypothetical scenario costs something, I'd rather not pay the price - especially for migrations.

**Comment by mercmobily** [comment]
@Bessonov Thanks for weighing in, but I think this thread is about something narrower: importing a dump for seeding/setup (dev/test parity, onboarding, CI fixtures), not broad cross-DB migration philosophy.

- Scope: The ask is a practical way to load a .sql dump with Knex. It's not trying to make migrations fully cross-DB or replace native tools.
- Pragmatism: Knex already embraces DB-specific needs via `raw()` in migrations/seeds. Supporting a dump import path fits that reality rather than changing Knex's philosophy.
- Single-DB benefit: Even teams that never plan to switch databases keep canonical dumps for fixtures. They'd benefit from a first-class or well-documented approach.
- Current workarounds hurt: `multipleStatements: true` has security and memory drawbacks; shelling out to the native client complicates CI/portability. That's why people want a better-supported pattern.
- Opt-in is fine: A DB-specific, opt-in recipe or plugin wouldn't force cross-DB purity on anyone, but it would solve a common pain.

Curious: with your "separate query builder instance" approach, have you seen concrete memory improvements for large dumps? If so, details could help shape a minimal, DB-scoped solution (doc recipe or plugin) that keeps this issue focused on the seeding use case.

**Comment by Bessonov** [comment]
@mercmobily 
Thanks for the clarification!

Sorry, I can't provide more details, because I haven't used Knex and MySQL for some time and have successfully avoided importing large dumps.

What I mean is to create separate instances of your tool for use within your app. For example (hypothetical code):
````ts
const myAppConnection = new Knex({ multipleStatements: false })

async function importDump(sql: string) {
  const myDumpConnection = new Knex({ multipleStatements: true })
  await myDumpConnection.execute(sql)
}
````

However, if the dumps are really big, you need some sort of streaming solution to read your migration files as a stream and pass that stream to the database. Never done it with knex.

**Comment by mercmobily** [comment]
Thanks for clarifying. A few follow-ups to keep this actionable for Knex users who hit the dump-import case:

- Separate instance helps isolate risky settings, but it doesn't solve the two main pain points: security (still only safe for trusted dumps) and memory (still an issue if you load the whole file into a string). Also note Knex exposes `raw`, not `execute`.
- For large dumps you really do need streaming: read the file as a stream, split into statements, and execute sequentially. Knex doesn't ship a splitter, so today the practical options are: (1) shell out to the native `mysql` client and pipe the file (works and is memory-efficient), or (2) implement a small statement-aware splitter that feeds `knex.raw` per statement with a `multipleStatements: true` connection that you dispose right after (`await knex.destroy()`).
- The "separate connection just for import" pattern is still useful: it limits the blast radius of `multipleStatements: true`, keeps different timeouts, and is easy to tear down. But to avoid leaks, either stream or hand off to the native client; otherwise big dumps will still bloat memory.
- If anyone has a robust streaming splitter pattern in Knex/Node they're happy with, that would be great to document. Short of that, a blessed "use the native client" recipe would cover the common case.

(Addendum: sorry for keeping this short)
